### PR TITLE
No shard lock on I/O: task request tracker

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Add task request tracker for shard context

**Why?**
- Pull out logic related to tracking pending task keys and inflight persistence requests. Needed for later work for releasing shard lock on shard I/O. For the full change check: https://github.com/temporalio/temporal/tree/remove-shard-lock-v2/service/history/shard

**How did you test it?**
- Unit test & tested in test cluster

**Potential risks**
- N/A, not wired up yet

**Is hotfix candidate?**

### CODE REVIEW

File: service/history/shard/task_request_tracker.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+				continue
+			}
+
+			if _, ok := minKeyByCategory[category]; !ok {
+				minKeyByCategory[category] = minKey
+			} else {
+				minKeyByCategory[category] = tasks.MinKey(minKeyByCategory[category], minKey)
+			}
+		}
+	}
```
Comment (tdeebswihart) [inline service/history/shard/task_request_tracker.go:87]: Do we need to hold the mutex while we do this? It looks like we could do this outside of the critical sections

Comment (yycptt) [reply service/history/shard/task_request_tracker.go:87]: 👍 Make sense. I will move the computation of `minKeyByCategory` out of the lock.

File: service/history/shard/task_request_tracker.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+)
+
+func newTaskRequestTracker() *taskRequestTracker {
+	outstandingTaskKeys := make(map[tasks.Category]map[tasks.Key]struct{})
+	for _, category := range tasks.GetCategories() {
+		outstandingTaskKeys[category] = make(map[tasks.Key]struct{})
+	}
+	return &taskRequestTracker{
+		pendingTaskKeys: outstandingTaskKeys,
+		waitChannels:    make([]chan<- struct{}, 0),
```
Comment (yux0) [inline service/history/shard/task_request_tracker.go:56]: any difference to init with 0 length and set it to nil at L170?

File: service/history/shard/task_request_tracker.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	}
+}
+
+func (t *taskRequestTracker) track(
+	taskMaps ...map[tasks.Category][]tasks.Task,
```
Comment (yux0) [inline service/history/shard/task_request_tracker.go:61]: trying to understand why we need to take in a list of maps

File: service/history/shard/task_request_tracker.go (hunk 2 above)
Comment (yycptt) [reply service/history/shard/task_request_tracker.go]: No I think. Not sure what I am thinking when writing this. Removed.

File: service/history/shard/task_request_tracker.go (hunk 3 above)
Comment (yycptt) [reply service/history/shard/task_request_tracker.go:60]: Synced offline. The request from shard context will be in this format and I don't want to incur unnecessary computation to flatten the structure.

**Review by yux0 — APPROVED** [review]

**Pull request merged**